kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/PartitionUtils.java

```diff
@@ -248,6 +251,44 @@ private static Expression rewritePartitionColumnRef(
         return expression;
     }
 
+    /**
+     * Get the target directory for writing data for given partition values.
+     *
+     * @param dataRoot          Root directory where the data is stored.
+     * @param partitionColNames Partition column names. We need this to create the target directory
+     *                          structure that is consistent levels of directories.
+     * @param partitionValues   Partition values to create the target directory.
+     * @return Target directory path.
+     */
+    public static String getTargetDirectory(
+            String dataRoot,
+            List<String> partitionColNames,
+            Map<String, Literal> partitionValues) {
+        Path targetDirectory = new Path(dataRoot);
+        for (String partitionColName : partitionColNames) {
+            Literal partitionValue = partitionValues.get(partitionColName);
+            checkArgument(
+                    partitionValue != null,
+                    "Partition column value is missing for column: " + partitionColName);
+            String serializedValue = serializePartitionValue(partitionValues.get(partitionColName));
+            if (serializedValue == null) {
+                //
+                serializedValue = "__HIVE_DEFAULT_PARTITION__";
```

Comment:
any docs, link you can give for this?

Reply:
This isn't documented in our protocol because this isn't enforced anywhere. The data can be any directory under the table root, and `AddFile` refers to the path. This is what delta-spark does to enable reading these tables in Hive. I don't know the exact historical context of why we do this, but my guess is to enable Hive reading it.